### DISCUSSION TITLE
Protocols and/or ports based passthrough policy

### DISCUSSION THREAD

**Discussion by andy-ning** [strongswan/strongswan#2354] [General]
Hi, I'll need to add some passthrough policy to bypass certain services by port numbers. The following example is from https://docs.strongswan.org/docs/6.0/config/quickstart.html#_for_specific_protocols_andor_ports

For specific protocols and/or ports
The following is a passthrough policy that allows traffic to the local SSH port from any remote address/port.
/etc/swanctl/swanctl.conf:
  connections {
    passthrough-ssh {
      **remote_addrs = 127.0.0.1**
      children {
        passthrough-ssh {
          local_ts  = dynamic[tcp/22]
          remote_ts = 0.0.0.0/0
          mode = pass
          start_action = trap
        }
      }
    }
  }

Two questions:
1. What is the "remote_addrs = 127.0.0.1" for? Why it is needed? 
2. What should remote_addrs and remote_ts be set to for IPv6 system? 

Thanks,
Andy

**Comment by tobiasbrunner** [comment]
> What is the "remote_addrs = 127.0.0.1" for? Why it is needed?

As documented on that page, it's to avoid that the connection is used as an actual IKE connection (no IKE packet will have a source address of `127.0.0.1`, so this connection will not be eligible when looking for a connection).


> What should remote_addrs and remote_ts be set to for IPv6 system?

Except for the discussion above, the value for `remote_addrs` doesn't really matter for shunt policies (drop/pass). `remote_ts` should obviously be one or more IPv6 address/subnet. Use `::/0` to match any IPv6 address, or use `0.0.0.0/0,::/0` for both address families (`dynamic` already works for both, actually, you could even set `remote_ts = dynamic` for the same effect).

**Reply by andy-ning** [reply]
Thanks. So my interpretation is that, for shunt policies (drop/pass), remote_addrs can be left unspecified. (ie, no remote_addrs defined for the connection). Does this sounds right? 

(I'd rather not specify remote_addrs to avoid confusion, because I'm not fully understand this statement on the page "remote_addrs is set to 127.0.0.1 to prevent this connection from being considered if a peer connects.")

**Reply by tobiasbrunner** [reply]
The problem is that these connection entries will then be considered when looking for a config for an inbound IKE message. The initial lookup is based on the IP addresses, the IKE version and the proposals. Since `remote_addrs` defaults to `%any`, the IKE version to 0 (i.e. any) and there is a comprehensive default proposal, the connection entry will basically match any inbound IKE message. There might later be a switch to a different entry (e.g. because the identities won't match), but there could also be an authentication failure (e.g. because a certificate is missing). The order in which connections are loaded does make a difference there, but just configuring `remote_addrs=127.0.0.1` avoids it altogether as the connection will be skipped whatever the order is.

**Reply by andy-ning** [reply]
Though I don't fully understand error conditions that can be run into (and I don't see them in my test in transport mode, bypassing SSH), I think it makes senses. By specifying remote_addrs=127.0.0.1 will prevent this connection from being matched for IKE packets all together, because no IKE packets will be from 127.0.0.1.

**Reply by tobiasbrunner** [reply]
You could increase the log level for _cfg_ to see more about the connection matching.

**Comment by andy-ning** [comment]
Hi, I've got another question on port range calculation. In the "For specific port ranges" example on page https://docs.strongswan.org/docs/6.0/config/quickstart.html#_for_specific_protocols_andor_ports, why the port range 65000-65255 is divided into 6 subranges? Would one range with 256 ports work, so that we can configure using "dynamic[tcp/65000-65255]"? 

I need to bypass port 6800 - 7300, how would I split it into subranges? Any documents talking about that?

Thanks.

**Reply by tobiasbrunner** [reply]
> why the port range 65000-65255 is divided into 6 subranges? Would one range with 256 ports work, so that we can configure using "dynamic[tcp/65000-65255]"?

No, port ranges have to be bitmasked in the Linux kernel (like subnets). There is no single mask to cover that range unless you want to include a lot of other ports too. That is, if you configured `65000-65255` you'd get a bitmask of `0xfc00`, which actually covers the range `64512-65535`.


> I need to bypass port 6800 - 7300, how would I split it into subranges? Any documents talking about that?

If you configured that, you'd get a bitmask of `0xf800`, which covers `6144-8191`. To make that smaller, you have to split your range into subranges that can be represented by a bitmask (i.e. a power of two).

Looking at `6800` in binary (`1101010010000`), the least significant bit set is the fifth, so the first range is `6800-6815` (mask `0xfff0`), the next range starts with `6816`, whose least significant bit set is sixth, so the next range is `6816-6847` (mask `0xffe0`), the next range is `6848-6911` (mask `0xffc0`), then `6912-7167` (mask `0xff00`). You can't continue in the same fashion as you don't want to overshoot `7300`, so you might want to proceed backwards. `7300` can't be covered by a range, so that will be a single port. Subtracting one, gets you the end of the next range, `7299`, which has the last two bits set (`1110010000011`). Flipping these gets you the range, `7296-7299` (mask `0xfffc`). Subtracting one gets you the end of the next range, `7295`, which has the last seven bits set, so the next range is `7168-7295` (mask `0xff80`) and we are done: `6800-6815`, `6816-6847`, `6848-6911`, `6912-7167`, `7168-7295`, `7296-7299`, `7300`.

**Reply by andy-ning** [reply]
Thank you very much! I kind of figured it out but your explanation is more thorough. This is very helpful. (I'd suggest include an example like this in the document.)
